### ISSUE TITLE
Unattended configuration stategy / application.ini / revert version

### ISSUE THREAD

**Issue by mmallejac** [the-djmaze/snappymail#1618]
I am trying to figure out what would be the best way to handle the modifications done on `application.ini` version after version, along with the settings we want to modify, considering that:
- we have many Nextcloud + Snappymail docker instances, so doing it manually is no fun
- we test new versions on one or two instances before deploying to the others
- we use bash scripts which use `crudini` to set values in `application.ini`

Last Snappymail upgrade (2.36.1 -> 2.36.2 / 2.36.3) has lead to unstable situation, most probably because `application.ini` file has been messed up: some values has turned from On to Off without obvious reasons (for instance `allow_additional_accounts`). 

I've seen that the `application.ini` file lives its life, with added, removed or moved items, version after version, that is fine.

- how is the `application.ini` file updated when a new Snappymail version is installed ?
- I've tried to find a vanilla version within the github repo, but it seems that the file is generated ?
- do you think we should set _all_ values in each section to be sure not to loose stuff like allow_additional_accounts ?

And an extra topic: do you know a way to revert from 2.36.3 to 2.36.1 ? We have backups, but so far Nextcloud does not allow us to do so (most probably the app version is somewhere in the Nextcloud database). Would you go that way to try to fix our broken installations ?

Thanks a lot for your help !

**Comment by mmallejac** [comment]
Update : downgrade Snappymail from 2.36.3 to 2.36.1 is doable:

- `occ app:disable snappymail`
- restore `apps/snappymail` (or `custom_apps/snappymail`) from backup
- restore `data/appdata_snappymail/` from backup
- `mysql` into the Nextcloud database then 
  - `select * from oc_appconfig where appid = "snappymail";` 
  - `update oc_appconfig set configvalue="2.36.1" where appid="snappymail" and configkey="installed_version";` 
- `occ app:enable snappymail`

Et voilà.

**Comment by the-djmaze** [comment]
Could this be an issue with crudini?
I have never seen issues with modifying using PHP scripts.

**Comment by mmallejac** [comment]
I don't think so since this script use to run every day in the past (without problems actually). 
But we have decided to no longer run it on a regular basis but only in few cases : starting a new instance or modifications we want to push to the existing instances.
Also we now test upgrades on one or two instances and wait until we see that all is ok before pushing to the other ones.

To be honest I am not exactly sure about the order of what I did when things have started to smell bad with 2.36.2 then 2.36.3. Probably at some point I've run my config / crudini script. Not sure at what point exactly the `application.ini` has been broken.

Now that I know more about all this I will have a much closer look with the future updates !

I now stick to the 2.36.1 and will wait the next one.

**Comment by mmallejac** [comment]
I re-open this, since I've just seen #1601 I guess they are experimenting the same thing !

**Comment by mmallejac** [comment]
I re-open this, since I've just seen #1601 I guess they are experimenting the same thing !

After reverting to 2.36.1 I've also updated all the plugins that had a new version.

**Comment by the-djmaze** [comment]
> * we have many Nextcloud + Snappymail docker instances

Aren't you using persistent storage?
Because now it sounds like you are moving things around in docker instances and then modify application.ini and it screws it up.

**Comment by mmallejac** [comment]
I have a persistent volume that holds the `/var/www/html` folder of the Nextcloud container.

That is in place for many time now and has gone threw quite a lot of Nextcloud updates (some of them from v18 to v29).

**Comment by the-djmaze** [comment]
So there is no need for your script because on SnappyMail updates everything should stay the same.

**Comment by the-djmaze** [comment]
Can you figure out why your script is corrupting the application.ini ?

**Comment by the-djmaze** [comment]
Can you figure out why your script is corrupting the application.ini ?

Also remove any rule that changes the `language` because there are `language` settings in the ini.

**Comment by mmallejac** [comment]
Thanks for digging into this !

I don't think the script has corrupt the application.ini, since the script has been run thousands of time in the past, and it has not been modified for some time.  The possibility exists, for sure, but I never had any signs of this kind of troubles in the past.

For the language, this is what I have in the script:
`crudini --set --verbose $snapCfg webmail language              "\"fr_FR\""`

I've seen that the `application.ini` was corrupt after 2.36.1 -> 2.36.2 -> 2.36.3
But not sure at what point exactly. 

I've started to see weird thing after 2.36.2, and shortly after 2.36.3 was available, I've pushed it but still had weird things.
Then I've restarted the instance and also run the script. No change, still weird things
Then I've started to investigate more deeply.

So I'd say that it was the 2.36.1 -> 2.26.2 update, but not really sure !
